### DISCUSSION TITLE
Spyglass

### DISCUSSION THREAD

**Discussion by Apprendista81** [mainsail-crew/spyglass#90] [Q&A] [answered]
Hi, 
![IMG_20240918_180205_383~2](https://github.com/user-attachments/assets/4b9173f6-a995-41cd-82f7-6b728f419feb)
![IMG_20240918_180205_383~2](https://github.com/user-attachments/assets/527c573e-26a1-43cc-86fe-21ee08ebd530)
I installed Spyglass on a Raspberry pizero2w to use the Raspberry cam through mainsail when I use the 3d printer but I can't view the video it gives me a gray screen with a small photo icon in the top left. What do I have to do to see the transmission? thank you very much in advance

**Comment by mryel00** [comment]
Send your spyglass.conf and the settings used in the Mainsail UI. Also run `journalctl -u spyglass > ~/printer_data/config/journalctl.txt` and download & send the journalctl.txt appearing on you machine page.

**Comment by Apprendista81** [comment]
I am not familiar with exposing in the forum what steps I have to do? I used your instructions to install it from the terminal, add string to monraker.conf and set the CAM setting strings. It gives me an error: spyglass.service: failed with result 'exit-code'

Inviato da Outlook per Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by mryel00** [reply]
> I am not familiar with exposing in the forum what steps I have to do?

Didn't I just tell you, what I need to know? Send the things I asked you for, then I can help you further.

**Reply by mryel00** [reply]
If you are wondering about data leakage, there shouldn't be anything you should have to worry about, unless you are using some public IP in the Mainsail UI settings, but then you are doing something horribly wrong anyway.

**Comment by Apprendista81** [comment]
I can't connect with the terminal from the PC to the Raspberry, it gives me this response :@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!
Someone could be eavesdropping on you right now (man-in-the-middle attack)!
It is also possible that a host key has just been changed.
The fingerprint for the ED25519 key sent by the remote host is
SHA256:5Kc8f3M7ssCLmuJbRyiuNuJp9duHqv+G1iWFfo6otq0.
Please contact your system administrator.
Add correct host key in C:\\Users\\Dovbi/.ssh/known_hosts to get rid of this message.
Offending ECDSA key in C:\\Users\\Dovbi/.ssh/known_hosts:35
Host key for 192.168.54.97 has changed and you have requested strict checking.
Host key verification failed.

**Reply by mryel00** [reply]
There are 2 possibilities:
- Someone is really doing a man-in-the-middle attack in your local network (then you got some more serious problem, very unlikely)
- You reflashed the Pi since the last time you connected through ssh (way mor likely)

To "fix" it, you could just remove all the entries with `192.168.54.97` in your `C:\Users\Dovbi/.ssh/known_hosts`. Then try it again. Only do this, if you are sure that you changed something on `192.168.54.97`, like reflashing, since the last time you connected via ssh to that device.

**Reply by Apprendista81** [reply]
I opened with notes and removed them, then how should I record the change?

**Reply by mryel00** [reply]
> then how should I record the change?

Sry, but I don't understand what you mean. Just try to connect again, after you deleted and saved it

**Reply by Apprendista81** [reply]
same message, doesn't work

**Reply by mryel00** [reply]
If it's still the same message, you didn't delete all the entries with that ip. It should be a different message, after deleting all entries corresponding to that ip

**Reply by Apprendista81** [reply]
I deleted the entire page, not just the IP, and reconnected as at the origin. Now I'll type the commands you wrote to me and copy the results

**Reply by Apprendista81** [reply]
pi@pi:~ $ ~/printer_data/config/journalctl.txt
-bash: /home/pi/printer_data/config/journalctl.txt: No such file or directory
pi@pi:~ $ ~/printer_data/config/journalctl.txt
-bash: /home/pi/printer_data/config/journalctl.txt: No such file or directory
pi@pi:~ $ journalctl -u spyglass > ~/printer_data/config/journalctl.txt
pi@pi:~ $

**Reply by mryel00** [reply]
Weird that you stroke all the text, but can you maybe elaborate more on your setup? I guess your pi zero 2w is separate to your Klipper+Mainsail Pi?

**Reply by Apprendista81** [reply]
https://pastebin.com/VtuDhEDs

**Reply by mryel00** [reply, marked as answer]
Oh, it's a Bookworm image. I didn't yet update the installation instructions.
Run the following commands:
```sh
sudo apt update
sudo apt install python3-libcamera python3-kms++ python3-picamera2 python3-v4l2 -y
```
After that reboot and that error should be gone

**Reply by Apprendista81** [reply]
in what sense is it separate? the raspberry is connected to the printer and now I use the desktop PC connected via ssh to the pi02w now or mainsail opened from the desktop PC

**Reply by mryel00** [reply]
The question came up because of the following message of yours
> pi@pi:~ $ ~/printer_data/config/journalctl.txt
-bash: /home/pi/printer_data/config/journalctl.txt: No such file or directory
pi@pi:~ $ ~/printer_data/config/journalctl.txt
-bash: /home/pi/printer_data/config/journalctl.txt: No such file or directory
pi@pi:~ $ journalctl -u spyglass > ~/printer_data/config/journalctl.txt
pi@pi:~ $

But I didn't read it correctly as it got crossed out weirdly.
Fyi after the last line it should have created a journalctl.txt that you can download on the Mainsail machine page btw

**Reply by Apprendista81** [reply]
> Oh, è un'immagine di Bookworm. Non ho ancora aggiornato le istruzioni di installazione. Esegui i seguenti comandi:
> 
> ```shell
> sudo apt update
> sudo apt install python3-libcamera python3-kms++ python3-picamera2 python3-v4l2 -y
> ```
> 
> Dopo il riavvio, l'errore dovrebbe essere scomparso

was updating

**Reply by Apprendista81** [reply]
hurray now you can see it, thank you very much

**Reply by Apprendista81** [reply]
yet I followed all the information initially, but I think I didn't realize that he hadn't updated it. However, I noticed that with raspberry image there is no longer the option to download mainsail directly with bookworm.
thank you thank you thank you

**Reply by Apprendista81** [reply]
How do I mark it as resolved?

**Reply by mryel00** [reply]
> yet I followed all the information initially, but I think I didn't realize that he hadn't updated it.

Those packages that you just installed don't get mentioned yet in the install section. That's something I still have to add.

> However, I noticed that with raspberry image there is no longer the option to download mainsail directly with bookworm

Mainsail is not MainsailOS. We (the mainsail crew) are currently still developing on the Bookworm image, as there are some things that changed and need some workarounds to get the same user experience as with Bullseye.

**Comment by Apprendista81** [comment]
if I can ask a question even if the problem has been solved, using a pi02w with bookworm and a cam connected, can spyglass be used as a surveillance camera or to be connected to a telescope?

**Comment by mryel00** [comment]
First of all I recommend to open a new post for a new question, just that people can find that question later on, if they got the same.
But I don't really understand the intention of this question. Spyglass only streams the pictures of a picam to the network (only local network, if you didn't setup something special on your own). So it cannot record anything without modification, if that's what you want to know. It also cannot effectively use usb cameras yet (that might be added in the future).

You can easily add a second output that it will output the stream to a file, in theory, but that's nothing I would consider adding to Spyglass myself, as the usecase is mainly meant for monitoring your 3D printer and in such a usecase you don't need recordings to a file, except for timelapses maybe, but I guess thi would get too complicated for Spyglass itself and should be it's own application then.

For the change I mentioned, it would be simply adding the following line to `spyglass/camera/csi.py` directly below line 33, containing similar code:
```python
self.picam2.start_recording(MJPEGEncoder(), FileOutput("path/to/output.mjpg"))
```
I didn't test it right now, so no guarantee that this should work. Also it's highly recommended to use H264 for recording as the compression is way better. But for more information, I recommend to open a new post, that it can be found.

**Reply by Apprendista81** [reply]
a real streaming even without recording to start, but I would need an app or something I can use to receive the signal and see it when needed, but with bookworm 64bit I don't know which one I could use.

**Reply by mryel00** [reply]
Overall Spyglass is just sending a mjpg stream, so in theory you can use everything that can watch a mjpg stream. I'm working on WebRTC currently too. But in the end, you can also just use Mainsail and e.g. Tailscale for remote access.

**Comment by Apprendista81** [comment]
I'm relatively new to all of this and have only used it on Klipper on the 3D printer. I should use it on my phone, what would you recommend? Are you right with written tutorials?

**Reply by mryel00** [reply]
I don't use any smartphone app and just use Mainsail in the browser on my phone. Mobileraker is a pretty common one, but both Mainsail and Mobileraker need Moonraker to be running afaik. Hence other apps to view streams, I don't know any, but maybe VLC or MPV can work, but I never tested it.
> Are you right with written tutorials?

I don't understand what you mean by that.

**Reply by Apprendista81** [reply]
I have only used mainsail from a desktop PC, I didn't know it was possible from a smartphone; I wrote it wrong, I was asking if you have a tutorial on how to set it all up, I'm very ignorant about this and little by little I'm trying to learn what can be useful.

**Reply by mryel00** [reply]
> I was asking if you have a tutorial on how to set it all up

What exactly do you mean by that?

**Reply by Apprendista81** [reply]
I downloaded vlc and mobilraker on my smartphone, now I have to turn on the Raspberry and how do I connect to the Raspberry with vlc without using mobilraker?

**Reply by Apprendista81** [reply]
I also downloaded tailscale on my smartphone

**Reply by mryel00** [reply]
> how do I connect to the Raspberry with vlc without using mobilraker

Please research on your own, on how to open a mjpg stream in VLC
